File: server/pkg/api/base_envs.go
package api

const (
	BaseEnvsQueryTypeByFactoryID = "base_envs/byFactory"
)

const (
	BaseEnvsTableName = "base_envs"
)

type BaseEnvsQueryByFactoryID struct {
	FactoryID int `json:"factoryId"`
}

type BaseEnvsTableRow struct {
	Recipes []BaseEnvsTableRowRecipe `json:"recipes"`

	TileFertility float64 `json:"tileFertility"`

	Resources  []WorldsTableRowResourceDeposit `json:"resources"`
	Snow       []WorldsTableRowResourceDeposit `json:"snow"`
	Ocean      []WorldsTableRowResourceDeposit `json:"ocean"`
	Atmosphere []WorldsTableRowResourceDeposit `json:"atmosphere"`
}

type BaseEnvsTableRowRecipe struct {
	RecipeID    string             `json:"recipeId"`
	EquipmentID string             `json:"equipmentId"`
	Inputs      map[string]float64 `json:"inputs"`
	Outputs     map[string]float64 `json:"outputs"`
}

File: ui/src/store/datafront/base_envs.ts
import { Inventory } from '@/domain/Inventory';
import type { Recipe } from '@/domain/Recipe';
import type { BaseEnvs } from '@/domain/Base';
import { createDatafrontTable } from '@/lib/datafront/table';
import {
    BaseEnvsQueryTypeByFactoryID,
    type BaseEnvsQueryByFactoryID,
    BaseEnvsTableName,
    type BaseEnvsTableRow,
} from '@/lib/net/types.generated';
import { ws } from '@/lib/net/ws';
import { updater, cleaner } from './misc';

export const dfBaseEnvs = createDatafrontTable<BaseEnvsTableRow, BaseEnvs>({
    name: BaseEnvsTableName,
    ws,
    updater,
    cleaner,
    map: (data) => {
        return {
            dynamicRecipes: data.recipes.map(
                (rData): Recipe => ({
                    id: rData.recipeId,
                    equipment: rData.equipmentId,
                    inputs: Inventory.from(rData.inputs),
                    outputs: Inventory.from(rData.outputs),
                }),
            ),
            tileConditions: {
                soilFertility: data.tileFertility,
                atmosphericResources: data.atmosphere,
                oceanResources: data.ocean,
                snowResources: data.snow,
                resources: data.resources,
            },
        };
    },
});

export const dfBaseEnvsByFactoryId = dfBaseEnvs.createQuery<BaseEnvsQueryByFactoryID>(
    BaseEnvsQueryTypeByFactoryID,
    (payload) => payload.factoryId.toString(),
);
